File: fibonacci-dotnet/src/Fibonacci.Api/Algorithms/NaiveRecursiveAlgorithm.cs
using System.Numerics;
using Fibonacci.Api.Constants;

namespace Fibonacci.Api.Algorithms;

/// <summary>
/// Naive recursive Fibonacci — O(2^n) time, O(n) stack space.
/// Educational: demonstrates the exponential cost of recomputing overlapping sub-problems.
/// Capped at MaxNNaive to prevent prohibitive computation times.
/// </summary>
public sealed class NaiveRecursiveAlgorithm : IFibonacciAlgorithm
{
    public string Name => "recursive";
    public string Label => "Naive Recursive";
    public string TimeComplexity => "O(2^n)";
    public string SpaceComplexity => "O(n)";
    public int MaxN => FibonacciConstants.MaxNNaive;
    public bool RequiresExplicitInclusion => true;

    public BigInteger Compute(int n)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(n);
        return Fib(n);
    }

    private static BigInteger Fib(int n) =>
        n <= 1 ? n : Fib(n - 1) + Fib(n - 2);
}

File: fibonacci-java/src/main/java/com/bretzelcoder/fibonacci/algorithm/NaiveRecursiveAlgorithm.java
package com.bretzelcoder.fibonacci.algorithm;

import com.bretzelcoder.fibonacci.config.FibonacciConstants;
import java.math.BigInteger;
import org.springframework.core.annotation.Order;
import org.springframework.stereotype.Component;

/**
 * Naive recursion — O(2^n) time, O(n) stack.
 * Kept for its educational value: it recomputes the same sub-problems over and over,
 * which is exactly what the memoized variant exists to avoid. Bounded at
 * {@link FibonacciConstants#MAX_N_NAIVE} because the cost past that is prohibitive.
 */
@Component
@Order(1)
public class NaiveRecursiveAlgorithm implements FibonacciAlgorithm {

    @Override
    public String name() {
        return "recursive";
    }

    @Override
    public String label() {
        return "Naive Recursive";
    }

    @Override
    public String timeComplexity() {
        return "O(2^n)";
    }

    @Override
    public String spaceComplexity() {
        return "O(n)";
    }

    @Override
    public int maxN() {
        return FibonacciConstants.MAX_N_NAIVE;
    }

    @Override
    public boolean requiresExplicitInclusion() {
        return true;
    }

    @Override
    public BigInteger compute(int n) {
        requireNonNegative(n);
        return fib(n);
    }

    private static BigInteger fib(int n) {
        return n <= 1
                ? BigInteger.valueOf(n)
                : fib(n - 1).add(fib(n - 2));
    }
}
